### PR TITLE
Add advanced fuzzer (#724)

### PR DESCRIPTION
**Pull request by xgreenx**
PR currently blocked on https://github.com/FuelLabs/fuel-vm/issues/815

Closes https://github.com/FuelLabs/fuel-core/issues/1443
Related https://github.com/FuelLabs/fuel-vm/issues/718

The [original PR](https://github.com/FuelLabs/fuel-vm/pull/724). I moved the change to our codebase to see how CI works, plus it allows us to commit changes as well. Huge thanks to the @maxammann for finalizing the PR!

Edit (@netrome): I have done a few modifications on the original PR to make this ready to merge.

- Cleaned up unused code and clippy errors so the PR passes CI.
- Removed the outdated and broken `grammar_aware` fuzz target.
- Updated the README.md with instructions for how to generate a corpus, run the fuzzer, generate code coverage reports.
- Added an example initial corpus so people don't have to generate a corpus from scratch to run the fuzzer.
- Added the `libafl` and `libfuzzer` feature flags to toggle between the LibAFL and LibFuzzer fuzzers, without having to manually comment out dependencies, as the original PR required.
- Did an initial investigation of discovered crashes, and created https://github.com/FuelLabs/fuel-vm/issues/815 to further investigate and resolve this.

### Before requesting review
- [x] I have reviewed the code myself
- [x] I have created follow-up issues caused by this PR and linked them here

### CODE REVIEW

File: fuel-vm/fuzz/README.md
Comment (AurelienFT) [inline fuel-vm/fuzz/README.md:65]:
```suggestion
However, we recommend adding a few flags to the command to improve fuzzing efficiency. First, we can add `--no-default-features --features libafl` to ensure we use the LibAFL fuzzer instead of the default libFuzzer. Secondly, we can set `--sanitizer none` to disable AddressSanitizer for a significant speed improvement, as we do not expect memory issues in a Rust program that does not use a significant amount of unsafe code. This has been confirmed by a ToB [cargo-geiger](https://github.com/rust-secure-code/cargo-geiger) analysis showed. It makes sense to leave AddressSanitizer turned on if we use more unsafe Rust in the future (either directly or through dependencies). Finally, the `-ignore_crashes=1 -ignore_timeouts=1 -ignore_ooms=1 -fork=7` flags are useful to ensure a smooth LibAFL experience utilizing 7 cores.
```

File: fuel-vm/fuzz/src/lib.rs
Comment (AurelienFT) [inline fuel-vm/fuzz/src/lib.rs:36]: Is it possible to replace this function by https://doc.rust-lang.org/std/primitive.slice.html#method.split ?

File: fuel-vm/fuzz/README.md
Comment (netrome) [reply fuel-vm/fuzz/README.md]: Thanks for spotting!

File: fuel-vm/src/util.rs
Comment (xgreenx) [inline fuel-vm/src/util.rs:469]:
```suggestion
                .into_iter()
                .collect();
```

File: fuel-vm/src/util.rs
Comment (xgreenx) [inline fuel-vm/src/util.rs:451]:
```suggestion
        pub fn setup_contract_bytes(
```

Comment (xgreenx) [reply fuel-vm/src/util.rs:451]: It is okay to have this function=)

File: fuel-vm/src/util.rs
Comment (xgreenx) [inline fuel-vm/src/util.rs:215]:
```suggestion
        pub fn start_script_bytes(
```

File: fuel-vm/fuzz/src/lib.rs
Comment (xgreenx) [inline fuel-vm/fuzz/src/lib.rs:118]: I'm not sure why do we need this=)

File: fuel-tx/src/transaction/validity.rs
Comment (xgreenx) [inline fuel-tx/src/transaction/validity.rs:120]: Hmm, I think we should always produce a normalized signatures. Otherwise, it is hard to use in production. Maybe @Dentosal has some ideas?

**Review by xgreenx — COMMENTED** [review]
I think we don't need to include example corpus in the `fuel-vm` repository. The CI action could do all steps to produce corpus.

@netrome The final goal what we try to achieve is to run this fuzzer on nightly bases. Do you plan not add action in this PR, or it follow up?

File: fuel-vm/fuzz/src/lib.rs
Comment (netrome) [reply fuel-vm/fuzz/src/lib.rs:118]: Oh missed this one, will remove :+1:

File: fuel-tx/src/transaction/validity.rs
Comment (netrome) [reply fuel-tx/src/transaction/validity.rs:120]: Yeah, I think this in practice is very unlikely to ever happen - but the fact that we easily hit this in fuzzing warrants further investigation. I plan to look deeper into this in the context of https://github.com/FuelLabs/fuel-vm/issues/815

**Comment by netrome** [comment]
> I think we don't need to include example corpus in the `fuel-vm` repository. The CI action could do all steps to produce corpus.
> 
> @netrome The final goal what we try to achieve is to run this fuzzer on nightly bases. Do you plan not add action in this PR, or it follow up?

Fair enough, I'll remove the example corpus. I plan to follow up with the CI work in the [ClusterFuzzLite PR](https://github.com/FuelLabs/fuel-vm/pull/727).

File: fuel-vm/src/util.rs
Comment (netrome) [reply fuel-vm/src/util.rs]: Updated in https://github.com/FuelLabs/fuel-vm/pull/733/commits/86f73591b00cdc297ef4905c4ebd4b6ee0dc7d94

File: fuel-vm/src/util.rs
Comment (netrome) [reply fuel-vm/src/util.rs]: Done in https://github.com/FuelLabs/fuel-vm/pull/733/commits/86f73591b00cdc297ef4905c4ebd4b6ee0dc7d94

File: fuel-vm/fuzz/src/lib.rs
Comment (netrome) [reply fuel-vm/fuzz/src/lib.rs]: Didn't seem to make a difference. I've removed it.

File: fuel-tx/src/transaction/validity.rs
Comment (xgreenx) [reply fuel-tx/src/transaction/validity.rs:120]: Could you please look into https://github.com/FuelLabs/fuel-vm/issues/815 next?=) It would be nice to remove `#[cfg(not(fuzzing))]` before merging this PR

Comment (netrome) [reply fuel-tx/src/transaction/validity.rs:120]: Sure thing, I'll look into it=) Putting this PR on hold in the meantime.

Comment (netrome) [reply fuel-tx/src/transaction/validity.rs]: Done in https://github.com/FuelLabs/fuel-vm/pull/733/commits/ed5f70c8b62a4a3d8bceb9513e43bb3370ab95ba. I've removed the `#[cfg(not(fuzzing))]` now, and the fuzzer doesn't fail on those places anymore. The fix was as simple as bumping the `secp256k1` crate version.

File: fuel-crypto/src/secp256/signature_format.rs
Comment (xgreenx) [inline fuel-crypto/src/secp256/signature_format.rs:59]:
```suggestion
```

**Review by xgreenx — COMMENTED** [review]
LGTM!=) I can't approve since I created this PR initially=D

**Comment by netrome** [comment]
> LGTM!=) I can't approve since I created this PR initially=D

Nice, I'll approve in your place=)

**Review by netrome — APPROVED** [review]

File: fuel-vm/fuzz/src/lib.rs
Comment (netrome) [reply fuel-vm/fuzz/src/lib.rs]: We need the clones, but otherwise this looks better.

File: fuel-vm/fuzz/src/bin/collect.rs
Comment (netrome) [reply fuel-vm/fuzz/src/bin/collect.rs]: Right, done in https://github.com/FuelLabs/fuel-vm/pull/733/commits/bc8dbf242b971f1e3f92c783a5fea7787503ad90

File: fuel-vm/fuzz/src/lib.rs
Comment (rymnc) [reply fuel-vm/fuzz/src/lib.rs]: nice, thanks

File: fuel-vm/fuzz/src/bin/seed.rs
Comment (netrome) [reply fuel-vm/fuzz/src/bin/seed.rs]: Done in https://github.com/FuelLabs/fuel-vm/pull/733/commits/28fc3aedc0bf017d8c58a18ebfa3e3bc5eec8f1c

File: fuel-vm/fuzz/src/lib.rs
Comment (netrome) [reply fuel-vm/fuzz/src/lib.rs]: Done in https://github.com/FuelLabs/fuel-vm/pull/733/commits/b7f0d0bfb3ddaeb0302e47bd7a79cdc6e0e6df20

**Review by rymnc — APPROVED** [review]

**Review by netrome — APPROVED** [review]

File: fuel-vm/fuzz/src/lib.rs
Comment (Dentosal) [inline fuel-vm/fuzz/src/lib.rs:85]: Uuf

**Review by Dentosal — APPROVED** [review]

File: fuel-vm/fuzz/src/lib.rs
Comment (netrome) [reply fuel-vm/fuzz/src/lib.rs:85]: Yeah I feel you :sweat_smile:

**Pull request merged**
